### ISSUE TITLE
Double click a point cloud on the display area causes CC to hang

### ISSUE THREAD

**Issue by zoukyle** [CloudCompare/CloudCompare#618]
## Current Behavior
Double click a point cloud on the display area causes CC to hang

## Expected Behavior
Based on the release log, the double click should set the point as the rotation center

## Steps to Reproduce (for bugs)
Open a point cloud in CC on Linux 16.04 (I installed Linux on Macbook Pro), open a point cloud and double click any point.

## Context/Comments
I ran sudo perf top -p <CC process id> and I saw a lot of:
i965_dri.so
which makes me think it's related to the graphic card driver.

## Your Environment
* OS & Version: Linux 16.04 installed on Macbook Pro
* CC Version: 2.9.1
* Graphics card: 00:02.0 VGA compatible controller: Intel Corporation Crystal Well Integrated Graphics Controller (rev 08)

**Comment by dgirardeau** [comment]
Hum, Intel GPUs are indeed poorly supported.

Do you have the same issue when manually a point (SHIFT + left click on a point) and then telling CC you don't want to use the octree? (this way the driver will be used instead).

If CC doesn't ask the question, make sure the 'Auto-compute octree for picking' option is properly set in the 'Display' options.

**Comment by zoukyle** [comment]
I turned off 'autocompute octree' because I found it caused the hang more often. I don't have the same issue when manually pick a point.
I tried to modified the source code to comment out all double click event, but double click still caused the issue. Don't know how to get around. I only upgraded Qt to 5.9.2 and the graphic card driver to the latest. Still couldn't fix the problem.

**Comment by dgirardeau** [comment]
Hum that's strange... Especially since double clicking shouldn't trigger anything anymore in your version?!

**Comment by zoukyle** [comment]
Yeah, but I'm not 100% sure I commented all of the events. Can you reproduce the issue on your side?
My case is a little weird because I installed Ubuntu 16.04 on Macbook pro. I used to use CC on Windows and Mac, both of them are fine in the same Macbook pro machine. (I have triple OS installed)

**Comment by dgirardeau** [comment]
Okay, I reproduced the issue with a huge cloud. Actually the standard picking process (when you left click on the screen) was applied on each click, even when the user was double-clicking. With a lot of displayed entities (or big ones), the picking is too slow and the double-click is not recognized as such.

I (hopefully) fixed this (c825392eb5f6411a47eccee912fbf75076f3ee01). Can you test this on your side? (you'll need to recompile CC).

**Comment by asmaloney** [comment]
(That should be c5e0e760094fde618190fe04389c656c27b97a82 - the one you linked is a minor cleanup I made.)

**Comment by dgirardeau** [comment]
Oops :D. Thanks!

**Comment by zoukyle** [comment]
Thanks a lot for your quick fix. I'll give it a try tomorrow and let you know.

**Comment by zoukyle** [comment]
I tried to build from the master, but the issue still exists. Did you test on your side? My point cloud is 300M.

**Comment by dgirardeau** [comment]
Yes. I tried with the same amount of points. Make sure you got the c5e0e760094fde618190fe04389c656c27b97a82 commit.

**Comment by zoukyle** [comment]
I got the commit you mentioned. My CC version shows v2.10.alpha. Weird. How can I provide more information for debugging?

**Comment by dgirardeau** [comment]
Maybe you could output some debug messages (in the Console) in a few methods of ccGLWindow:
- ```ccLog::Warning("Button pressed");``` at the beginning of ccGLWindow::mousePressEvent
- ```ccLog::Warning("Double click");``` at the beginning of ccGLWindow::mouseDoubleClickEvent
- ```ccLog::Warning("Button released");``` at the beginning of ccGLWindow::mouseReleaseEvent
- ```ccLog::Warning("Picking");``` at the beginning of ccGLWindow::doPicking

Then start the program (in release mode preferably ;) and look at what happens. Ideally you should see something like:
- Button pressed
- Button released
- Double click
- Button released

And no 'Picking' message...

**Comment by tpwrules** [comment]
I have this issue as well with the same graphics driver. I found that it doesn't actually hang, but sits at 100% CPU usage for a few minutes. I can create hangs for a few seconds on even tiny clouds (~39K points).

I also found that I can set a rotation point if I double click fast. But double clicking slowly, or clicking more than twice, will definitely hang. Changing the "Auto-compute octree for picking" setting doesn't change the behavior. I don't get any indication that it's trying to pick a point in any case (e.g. setting "Always ask" never asks).

The new "iris" Intel driver in the latest Mesa also fixes the problem. Unfortunately, the snap comes with an older version.
